controllers/drpolicy_controller.go

```diff
@@ -98,66 +104,134 @@ func (r *DRPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
-			log.Info(`invalid -> invalid`)
-
-			return err
-		}
-		conditionSetTrue = func(reason, message string) error {
-			log.Info(`invalid -> valid`)
+			return b.statusConditionUpdate(condition, metav1.ConditionTrue, reason, message)
+		},
+		func() error {
+			b.log.Info("empty -> valid")
 
-			return conditionUpdate(metav1.ConditionTrue, reason, message)
-		}
-	} else {
-		conditionAppend := func(status metav1.ConditionStatus, reason, message string) error {
-			util.ConditionAppend(drpolicy, &drpolicy.Status.Conditions, ramen.DRPolicyValidated, status, reason, message)
+			return b.statusConditionAppend(ramen.DRPolicyValidated, metav1.ConditionTrue, reason, message)
+		},
+	)
+}
 
-			return client.Status().Update(ctx, drpolicy)
-		}
-		conditionSetFalse = func(reason string, err error) error {
-			log.Info(`empty -> invalid`)
-			if err1 := conditionAppend(metav1.ConditionFalse, reason, err.Error()); err1 != nil {
-				err = err1
-			}
+func (b *bar) validatedSetFalse(reason string, err error) error {
+	return b.validatedSet(
+		func(condition *metav1.Condition) error {
+			b.log.Info("valid -> invalid")
+
+			return error1stUnless2ndNotNil(
+				err,
+				b.statusConditionUpdate(condition, metav1.ConditionTrue, reason, err.Error()),
+			)
+		},
+		func(condition *metav1.Condition) error {
+			return b.statusConditionUpdateIfReasonOrMessageDiffers(condition, reason, err, "invalid -> invalid")
+		},
+		func() error {
+			message := err.Error()
+			b.log.Info("empty -> invalid", "reason", reason, "message", message)
+
+			return error1stUnless2ndNotNil(
+				err,
+				b.statusConditionAppend(ramen.DRPolicyValidated, metav1.ConditionFalse, reason, message),
+			)
+		},
+	)
+}
 
-			return err
+func (b *bar) validatedSet(foo1, foo2 func(*metav1.Condition) error, foo3 func() error) error {
+	if condition := util.DrpolicyValidatedConditionGet(b.drpolicy); condition != nil {
+		if condition.Status == metav1.ConditionTrue {
+			return foo1(condition)
 		}
-		conditionSetTrue = func(reason, message string) error {
-			log.Info(`empty -> valid`)
 
-			return conditionAppend(metav1.ConditionTrue, reason, message)
-		}
+		return foo2(condition)
 	}
 
+	return foo3()
+}
+
+func validate(ctx context.Context, drpolicy *ramen.DRPolicy, apiReader client.Reader,
+	objectStoreGetter ObjectStoreGetter, listKeyPrefix string,
+) (string, error) {
 	for i := range drpolicy.Spec.DRClusterSet {
 		cluster := drpolicy.Spec.DRClusterSet[i]
 		if reason, err := validateS3Profile(ctx, apiReader, objectStoreGetter,
 			cluster.S3ProfileName, listKeyPrefix); err != nil {
-			return conditionSetFalse(reason, err)
+			return reason, err
 		}
 	}
 
-	return conditionSetTrue(`Succeeded`, `drpolicy validated`)
+	return "", nil
+}
+
+func (b *bar) statusConditionUpdateIfReasonOrMessageDiffers(
+	condition *metav1.Condition,
+	reason string,
+	err error,
+	logMessage string,
+) error {
+	message := err.Error()
+	if condition.Reason == reason && condition.Message == message {
```

Comment:
If condition generation is older than resource generation, update status.

Reply:
I presume that applies even if the status, message, or reason do not change?

Reply:
That's true.

Reply:
commit https://github.com/RamenDR/ramen/pull/328/commits/7d256f925b4c25ce150b8eb9bf1bd3f733a39075 updates status condition if object's generation changes